20221012-add-type-restrictions-to-dsl-syntax.md

```diff
@@ -0,0 +1,117 @@
+ 2. Suppose A is an authorization model with type document and relation reader on document. We may have a type restriction on A that takes (document, reader) to the set {(user, \_), (group, member)}, meaning that the only users that may be readers of a document are users with type user, or a userset of group members.
+
+
+### Syntax
+
+Let A be an authorization model with set of types T and set of relations R, and recall that R* is the set R with the addition of an "empty" relation. We will write an element (t, r) of TxR* as t#r. The special case (t, _) is simply written as t. If the type restriction takes (t, r) to a set {(t1, r1), (t2, r2)}, then in our model we will write this as
+```
+type t
+    relations
+        define r: [t1#r1, t2#r2] as ...
+```
+If the type restriction takes (t, r) to the empty set we will simply write nothing in our model.
+
+Notice the the addition of type restrictions means we may drop `self` from the DSL. This is because if a pair (type, relation) has a non-empty set of type restrictions, then the assumption is that (type, relation) is a pair that allows a direct relation to type. This may become clear from the example below.
+
+#### Example
+
+Let's look at a concrete example. Consider the model:
+```
+type group
+    relations
+        define parent as self
+        define member as self
+        define guest as self
+        define can_view as guest
+```
+Suppose we want the following restrictions:
+1. parents of groups are groups
+2. members of groups are employees or group members
+3. guests of groups are users, employees, or group members
+
+We can write the model with these type restriction as
+```
+type user
+
+type employee
+
+type group
+    relations
+        define parent: [group]
+        define member: [employee, group#member]
+        define guest: [user, employee, group#member]
+        define can_view as guest
+```
+
+Notice that we no longer need `self` to indicate that a relation allows a direct relation.
+
+## Affect of type restrictions: writing tuples
+
+Given a model with type restrictions, we can now only write tuples that respect the type restriction. For example, given our model above the following are valid tuples:
+- object=group:X, relation=parent, user=group:Y
+- object=group:X, relation=member, user=employee:susan
+- object=group:X, relation=member, user=group:Y#member
+- object=group:X, relation=guest, user=user:susan
+- object=group:X, relation=guest, user=employee:mary
+- object=group:X, relation=guest, user=group:Y#member
+
+The following tuples would _not_ be valid:
+- object=group:X, relation=parent, user=frankie       # group parents can only be groups
+- object=group:X, relation=parent, user=user:frankie  # group parents can only be groups
+- object=group:X, relation=guest, user=group:Y        # group guests can only be group members, and not just groups
+
+
```

Comment:
For backward compatibility purpose, do we want to specify the schema version at the top of the DSL

For example
```
model
  schema 1.1
type user
type org
  relations
    define member: [user]
```

Reply:
Added. Thank you!